### PR TITLE
Training tweaks and loss wrapper

### PR DESCRIPTION
**Pull request by JLrumberger**
**What is the purpose of this PR?**

This PR closes Issue #21 by adding grokking training scheme, selective masking of the loss and a wrapper for loss functions from `tf.keras.losses` that allows to use label-smoothing.  

**How did you implement your changes**

Grokking training scheme:
- added weight decay to model_builder.py
- added cosine annealing learning rate scheduler

Loss masking and label smoothing:
- added a loss wrapper in loss.py that implements selective masking and uses tf.keras.losses loss functions instead of the deepcell ones. The keras losses allow label smoothing by default. 

**Remaining issues**

None

### CODE REVIEW

File: cell_classification/loss.py [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+            loss_img (tf.Tensor):
+                loss image
+            y_true (tf.Tensor):
+                ground truth image
+        Returns:
+            tf.Tensor:
+                masked loss image
+        """
+        y_true = tf.reshape(y_true, tf.shape(loss_img))
+        return tf.where(y_true == 2, tf.zeros_like(loss_img), loss_img)
```
Comment (ngreenwald) [inline cell_classification/loss.py:35]: how come we're checking if `y_true==2`?

File: cell_classification/configs/params.toml [hunk 2]
```diff
@@ -26,5 +26,12 @@ contrast_prob = 0.5
 contrast_min = 0.8
 contrast_max = 1.2
 batch_size = 4
+steps_per_epoch = 100
+loss_fn = "BinaryCrossentropy"
+loss_selective_masking = true
+[loss_kwargs]
+from_logits = true
+label_smoothing = 0.1
```
Comment (ngreenwald) [inline cell_classification/configs/params.toml:34]: It doesn't look like label_smoothing is ever popped from the params. Is this handled automatically somewhere?

Comment (ngreenwald) [reply cell_classification/configs/params.toml:34]: I assume that with only two classes, this won't have a huge impact, but given how easy it is to implement probably worth including?

**Review by ngreenwald — CHANGES_REQUESTED** [review]

File: cell_classification/loss.py (hunk 1 above)
Comment (JLrumberger) [reply cell_classification/loss.py:35]: I replaced the -1 for unspecific markers with 2 in order to store the groundtruth mask in uint8 instead of int32.

File: cell_classification/model_builder.py [hunk 3]
```diff
@@ -128,19 +138,17 @@ def train(self):
...
-        if n_classes == 1:
-            def loss_fn(y_true, y_pred):
-                return tf.keras.losses.binary_crossentropy(y_true, y_pred, from_logits=False)
-        elif n_classes > 1:
-            def loss_fn(y_true, y_pred):
-                return tf.keras.losses.categorical_crossentropy(y_true, y_pred, from_logits=False)
+        loss_fn = Loss(
+            self.params["loss_fn"], self.params["loss_selective_masking"],
+            **self.params["loss_kwargs"]
+        )
```
Comment (JLrumberger) [inline cell_classification/model_builder.py:151]: Here we pop params['loss_kwargs]

File: cell_classification/configs/params.toml (hunk 2 above)
Comment (JLrumberger) [reply cell_classification/configs/params.toml:34]: Not sure if I understand this correctly. So under `[loss_kwargs]` you can put all arguments that you want to put into the init of your tf.keras.loss. In [model_builder.py](https://github.com/angelolab/cell_classification/blob/8b7816cbb8fe041c34c40402d9949e73f6705159/cell_classification/model_builder.py#L150)  we pop `loss_kwargs` from params and it is used by the keras loss function to modify loss calculation.

**Review by ngreenwald — APPROVED** [review]
